src/sys/stat.rs

```diff
@@ -122,3 +122,48 @@ pub fn fstatat<P: ?Sized + NixPath>(dirfd: RawFd, pathname: &P, f: AtFlags) -> R
     Ok(dst)
 }
 
+/// Change the file permission bits of the file named by the pathname.
+///
+/// # References
+///
+/// [chmod(2)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/chmod.html).
+pub fn chmod<P: ?Sized + NixPath>(path: &P, mode: Mode) -> Result<()> {
+    let res =
+        path.with_nix_path(|cstr| unsafe { libc::chmod(cstr.as_ptr(), mode.bits() as mode_t) })?;
+
+    Errno::result(res).map(|_| ())
+}
+
+/// Change the file permission bits of the file specified by the file descriptor.
+///
+/// # References
+///
+/// [fchmod(2)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fchmod.html).
+pub fn fchmod(fd: RawFd, mode: Mode) -> Result<()> {
+    let res = unsafe { libc::fchmod(fd, mode.bits() as mode_t) };
+
+    Errno::result(res).map(|_| ())
+}
+
+/// Change the file permission bits.
+///
+/// # References
+///
+/// [fchmodat(2)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fchmodat.html).
+pub fn fchmodat<P: ?Sized + NixPath>(
+    dirfd: RawFd,
+    path: &P,
+    mode: Mode,
+    flag: AtFlags,
+) -> Result<()> {
+    let res = path.with_nix_path(|cstr| unsafe {
```

Comment:
This function actually has 3 modes:
 * Path relative to the cwd of the calling process
 * Path relative to the specified directory
 * Absolute path

I'd therefore suggest replacing `dirfd` and `path` with the following:
```rust
pub enum PathMode {
  RelPath(P),
  RelPath(P, dirfd),
  AbsPath(P),
}```

Then internally we can do the whole `AT_FDCWD` dance and also `debug_assert!` that those paths are correctly relative or absolute.

Reply:
There's no point in providing both `chmod` and `fchmodat(PathMode::AbsPath)`; they're identical.  Also, I don't think we should force a programmatic distinction between absolute and relative paths, because there are likely to be situations where a Nix consumer doesn't know which type of path it's dealing with, without inspecting it.  How about we provide `chmod`, `fchmod`, and `fchmodat`, where the latter asserts that `fd` is nonnegative?

Reply:
> asserts that fd is nonnegative

We can not because `AT_FDCWD` is negative integer usually.

Reply:
I'm wondering if we can't wrap up these 3 functions and 5 modes into a different object that provides a nicer API than these functions separately? There is definite overlap between these functions, and I don't like the terrible combinatorics of the inputs to `fchmodat`. Any ideas here?

Reply:
I think we shouldn't merge `fchmod` and `chmod`/`fchmodat` because these functions are very different (path vs fd as main argument).
We can get rid of `chmod` as it duplicates `fchmodat` and leave `fchmodat` with the signature:
```
pub fn fchmodat<P: ?Sized + NixPath>(path: P, mode: Mode, dirfd: Option<RawFd>, nofollow_symlink: bool) -> Result<()>
```
Here `None` in `dirfd` substitutes `AT_FDCWD` constant.

Reply:
I like that proposal.  I'd like it even better if it had the same argument order as `libc::fchmodat`.

Reply:
I highly suggest removing that boolean option. It makes reading the code much harder because you don't know what `True` and `False` actually mean without jumping into the docs. This is explicitly mentioned in [the Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-custom-type).

Reply:
Ok, we can replace the bool argument with:
```
pub enum FchmodatFlag {
    FollowSymlink,
    NoFollowSymlink,
}
```

Reply:
@asomers, @Susurrus The following is a summary of proposed changes:
1. Get rid of `chmod` function.
2. Add `FchmodatFlag` enum instead of `AtFlags`.
3. Change a signature of `fchmodat` function to: `pub fn fchmodat<P: ?Sized + NixPath>(dirfd: Option<RawFd>, path: P, mode: Mode, flag: FchmodatFlag) -> Result<()>`.

Do you have any objections or/and proposals?

Reply:
Yes, that sounds good to me.

Reply:
@Susurrus Any response?

Reply:
@antage @Susurrus is AFK.  Go ahead and make the changes we discussed, then I can approve the PR.

Reply:
@asomers ok.